File: packages/sdk/src/domains/games.ts
/**
 * The games domain: browse the catalogue, start a game, submit game actions, and observe game
 * state changes.
 *
 * Games are room-scoped and server-authoritative. A client never computes an outcome; it submits an
 * *intent* ({@link GameAction}) — "I play this card", "I roll" — and the server decides what happens
 * and broadcasts the result as a {@link GameEvent}. This split is deliberate (brief section 89): a
 * client cannot fabricate a win, because the client that submits the action is not the party that
 * rules on it. Rewards are virtual and non-monetary throughout (sections 37 and 87); there is no
 * real-money stake anywhere in this path.
 *
 * # Actions are deduplicated, events are deltas
 *
 * Each action carries an `actionId` that is monotonic per game for this player, so a resubmitted
 * action (a retry after a flaky connection) is recognised as the same one and not applied twice. This
 * domain mints that id from an in-memory per-game counter, so ordinary use never has to manage it.
 *
 * A {@link GameEvent} is a delta, never a full snapshot: it names an `event` and an optional binary
 * `payload` describing the change, plus a `stateVersion` a client uses to detect a missed event and
 * resync. A pre-rendered `text` line is included for thin clients that render the game as chat rather
 * than interpreting the payload.
 *
 * # Why a move's result needs a second read
 *
 * `GAME_ACTION`'s reply is a bare ack and the published events say only *that* somebody moved, so
 * the substance of a move — feedback on a guess, a board — is fetched with {@link getView} after the
 * ack resolves. {@link startGame} answers with the full opening view directly, because a game that
 * has not started has no deltas to publish.
 */

import type { Id } from '@migo/wire';
import {
  OP,
  encodeGameAction,
  decodeGameEvent,
  decodeAcknowledged,
  encodeGameStart,
  decodeGameViewWire,
  encodeGameId,
  encodeGiftCatalogueReq,
  decodeGameCatalogueResponse,
} from '@migo/protocol';
import type {
  Acknowledged,
  GameAction,
  GameCatalogueEntry,
  GameEvent,
  GameId,
  GiftCatalogueReq,
  GameStart,
  GameViewWire,
} from '@migo/protocol';

import { ListenerSet } from './listeners.js';
import type { Listener } from './listeners.js';
import type { EventErrorHandler, Rpc } from './rpc.js';

/**
 * Submit game actions and observe game events.
 *
 * One instance per client. {@link start} begins delivering events; {@link submit} works independently.
 */
export class GamesDomain {
  readonly #rpc: Rpc;
  readonly #listeners: ListenerSet<GameEvent>;
  /** The next action id to mint per game, so retries reuse an id and the server dedupes them. */
  readonly #nextActionId = new Map<Id, number>();
  #unsubscribe: (() => void) | null = null;

  constructor(rpc: Rpc, onEventError?: EventErrorHandler) {
    this.#rpc = rpc;
    this.#listeners = new ListenerSet(OP.GAME_EVENT, onEventError);
  }

  /** Begins delivering inbound game events to registered handlers. Idempotent. */
  start(): void {
    if (this.#unsubscribe !== null) {
      return;
    }
    this.#unsubscribe = this.#rpc.on(OP.GAME_EVENT, decodeGameEvent, (event) =>
      this.#listeners.deliver(event),
    );
  }

  /** Stops delivering game events. Registered handlers are kept for a later {@link start}. */
  stop(): void {
    this.#unsubscribe?.();
    this.#unsubscribe = null;
  }

  /** Registers a handler for inbound game events. Returns an unsubscribe function. */
  onGameEvent(handler: Listener<GameEvent>): () => void {
    return this.#listeners.add(handler);
  }

  /**
   * Reads the node's game catalogue: one entry per kind it can referee.
   *
   * The catalogue is the node's own and versionless — the same posture as the gift catalogue — so
   * a client re-reads it to build its menu each session rather than caching it. Each entry carries
   * the slug {@link startGame} accepts and the player counts a client needs to know which games it
   * can even offer.
   */
  async getCatalogue(): Promise<GameCatalogueEntry[]> {
    const request: GiftCatalogueReq = {};
    const response = await this.#rpc.call(
      OP.GAME_CATALOGUE,
      encodeGiftCatalogueReq,
      decodeGameCatalogueResponse,
      request,
    );
    return response.games;
  }

  /**
   * Starts a game in a conversation and resolves with the opening view.
   *
   * `slug` is a catalogue entry's slug. The wire names no opponents, so in this build a start can
   * open the single-player guessing game and nothing else — the server refuses a multi-player kind
   * with "wrong number of players" rather than inventing an opponent on the caller's behalf, and
   * that refusal is surfaced here as a {@link RemoteError}. Nothing is published to the
   * conversation on start: the reply carries the opening view to the caller, and the other members
   * hear of the game when its first move publishes a {@link GameEvent}.
   */
  async startGame(conversationId: Id, slug: string): Promise<GameViewWire> {
    const request: GameStart = { conversationId, slug };
    return this.#rpc.call(OP.GAME_START, encodeGameStart, decodeGameViewWire, request);
  }

  /**
   * Reads one game's current view, as the caller is allowed to see it.
   *
   * The view is redacted per viewer by the server, so this is also how a player learns the outcome
   * of their own move: {@link submit}'s reply is a bare ack, and a move's substance — a guess's
   * higher/lower, a board — lives in the fresh view, not in the published events, which say only
   * *that* somebody moved.
   */
  async getView(gameId: Id): Promise<GameViewWire> {
    const request: GameId = { gameId };
    return this.#rpc.call(OP.GAME_VIEW, encodeGameId, decodeGameViewWire, request);
  }

  /**
   * Submits a game action as an intent; the server decides and broadcasts the outcome.
   *
   * The `actionId` is minted automatically and monotonically per game, so submitting the "same" action
   * twice (a deliberate retry) would carry the next id and be treated as a new action — for an
   * idempotent retry, pass the previous id explicitly as {@link SubmitOptions.actionId}. Resolves once
   * the server has accepted the intent; the resulting state change arrives separately as a {@link
   * GameEvent}.
   */
  async submit(
    gameId: Id,
    roomId: Id,
    action: string,
    options: SubmitOptions = {},
  ): Promise<Acknowledged> {
    const actionId = options.actionId ?? this.#mintActionId(gameId);
    const request: GameAction = { gameId, roomId, actionId, action };
    if (options.args !== undefined) {
      request.args = options.args;
    }
    return this.#rpc.call(OP.GAME_ACTION, encodeGameAction, decodeAcknowledged, request);
  }

  /** Returns the next monotonic action id for a game, advancing the per-game counter. */
  #mintActionId(gameId: Id): number {
    const next = this.#nextActionId.get(gameId) ?? 1;
    this.#nextActionId.set(gameId, next + 1);
    return next;
  }
}

/** Optional parameters for {@link GamesDomain.submit}. */
export interface SubmitOptions {
  /** Arguments for the action, e.g. the coordinates of a move. */
  args?: string[];
  /**
   * An explicit action id, overriding the auto-minted one.
   *
   * Pass the id of a previous submission to retry it idempotently — the server recognises the repeated
   * id and does not apply the action a second time.
   */
  actionId?: number;
}

File: clients/android/core/src/main/kotlin/com/migo/core/domain/Games.kt
package com.migo.core.domain

import com.migo.core.protocol.Acknowledged
import com.migo.core.protocol.GameAction
import com.migo.core.protocol.GameCatalogueEntry
import com.migo.core.protocol.GameCatalogueResponse
import com.migo.core.protocol.GameEvent
import com.migo.core.protocol.GameId
import com.migo.core.protocol.GameStart
import com.migo.core.protocol.GameViewWire
import com.migo.core.protocol.GiftCatalogueReq
import com.migo.core.protocol.Op
import com.migo.core.wire.Id
import java.util.concurrent.ConcurrentHashMap

/**
 * In-room games: submitting a move, and receiving the authoritative result.
 *
 * A port of `packages/sdk/src/domains/games.ts`. The one domain whose model is deliberately
 * server-authoritative rather than peer-to-peer.
 *
 * # The server owns the game state, on purpose
 *
 * A client sends an *intent* ([submit]) and learns the *outcome* ([onEvent]). It never computes the
 * new state itself and never trusts a peer's claim about it. That is the only arrangement that
 * survives a player who edits their client: a game whose rules ran on the participants' devices is a
 * game whose rules are whatever the most modified device says they are.
 *
 * This is also why nothing here is encrypted. A game runs in a room, the room is not end-to-end
 * encrypted (see [RoomsDomain]), and a referee cannot referee a game it cannot see. The line is that
 * private messages are sealed and public game state is not, and conflating the two would mean either a
 * server that reads private chats or a game nobody can adjudicate.
 *
 * # Ordering comes from `stateVersion`, not from arrival
 *
 * [GameEvent.stateVersion] increments once per accepted action. A client renders in that order and
 * treats a gap as a missed event -- refetch the state rather than guessing at the intermediate steps.
 * A client that animated purely on arrival order would show two players' moves swapped whenever the
 * network reordered them, which in a game is not a cosmetic problem.
 *
 * # Why an action id exists, and why this class mints it
 *
 * [GameAction.actionId] makes a submission idempotent: the server records the id and answers a repeat
 * with the original outcome instead of applying the move twice. Without it, a reconnect during a
 * submit would be a coin flip between a lost move and a doubled one, and in a turn-based game a
 * doubled move is a lost game.
 *
 * The ids are minted here rather than by the caller because they must be *per game* and monotonic, and
 * a caller counting them would be a caller that resets the counter when it recreates its game object.
 * They live in memory only, which is the right lifetime: the server scopes them to a game session, so
 * a fresh process starting again at 1 is correct, not a collision.
 *
 * For a deliberate retry of a submission whose answer was lost, pass the [submit] `actionId` back
 * explicitly -- that is the whole point of the field, and a retry that let a fresh id be minted would
 * be a second move.
 */
class GamesDomain(
    private val rpc: Rpc,
    onEventError: EventErrorHandler? = null,
) {
    private val listeners = ListenerSet<GameEvent>(Op.GAME_EVENT, onEventError)

    /**
     * The next action id per game, keyed by the game id's text.
     *
     * A [ConcurrentHashMap] rather than a plain map behind a lock: [ConcurrentHashMap.compute] is
     * atomic per key, which is exactly the scope of the invariant -- two coroutines submitting for the
     * same game must not be handed the same id, and two submitting for different games have no reason
     * to contend. The reference uses a plain `Map` because JavaScript has one thread; Kotlin does not
     * get to make that assumption.
     *
     * Keyed by [Id.value] rather than by [Id] because a value class is boxed the moment it becomes a
     * generic type argument, and the box would be allocated on every lookup for no gain.
     */
    private val nextActionId = ConcurrentHashMap<String, Long>()

    @Volatile
    private var subscription: Subscription? = null

    /** Begins delivering game events to registered handlers. Idempotent. */
    fun start() {
        if (subscription != null) return
        subscription = rpc.on(Op.GAME_EVENT, { r -> GameEvent.decode(r) }) { event, _ ->
            listeners.deliver(event)
        }
    }

    /** Stops delivery. Registered handlers and the action-id counters are kept for a later [start]. */
    fun stop() {
        val live = subscription ?: return
        subscription = null
        live.cancel()
    }

    /**
     * Registers a handler for authoritative game events.
     *
     * [GameEvent.payload] is a game-specific byte string this SDK does not interpret; [GameEvent.text]
     * is an optional line the server composed for a client that has no renderer for this game, so a
     * generic room UI can still show something meaningful.
     */
    fun onEvent(listener: Listener<GameEvent>): Subscription = listeners.add(listener)

    /**
     * Reads the node's game catalogue: one entry per kind it can referee.
     *
     * The catalogue is the node's own and versionless — the same posture as the gift catalogue — so
     * a client re-reads it to build its menu each session rather than caching it. Each entry carries
     * the slug [startGame] accepts and the player counts a client needs to know which games it can
     * even offer.
     */
    suspend fun getCatalogue(): List<GameCatalogueEntry> {
        val request = GiftCatalogueReq()
        val response = rpc.call(
            Op.GAME_CATALOGUE,
            { w -> request.encode(w) },
            { r -> GameCatalogueResponse.decode(r) },
        )
        return response.games
    }

    /**
     * Starts a game in a conversation and resolves with the opening view.
     *
     * `slug` is a catalogue entry's slug. The wire names no opponents, so in this build a start can
     * open the single-player guessing game and nothing else — the server refuses a multi-player kind
     * with "wrong number of players" rather than inventing an opponent on the caller's behalf, and
     * that refusal surfaces here as a [com.migo.core.wire.WireError]. Nothing is published to the
     * conversation on start: the reply carries the opening view to the caller alone, and the other
     * members hear of the game when its first move publishes a [GameEvent].
     */
    suspend fun startGame(conversationId: Id, slug: String): GameViewWire {
        val request = GameStart(conversationId, slug)
        return rpc.call(
            Op.GAME_START,
            { w -> request.encode(w) },
            { r -> GameViewWire.decode(r) },
        )
    }

    /**
     * Reads one game's current view, as the caller is allowed to see it.
     *
     * The view is redacted per viewer by the server, so this is also how a player learns the outcome
     * of their own move: [submit]'s reply is a bare ack, and a move's substance — a guess's
     * higher/lower, a board — lives in the fresh view, not in the published events, which say only
     * *that* somebody moved.
     */
    suspend fun getView(gameId: Id): GameViewWire {
        val request = GameId(gameId)
        return rpc.call(
            Op.GAME_VIEW,
            { w -> request.encode(w) },
            { r -> GameViewWire.decode(r) },
        )
    }

    /**
     * Submits an action to a game.
     *
     * The returned [Acknowledged] says the server accepted the submission, not that the move
     * succeeded: the outcome arrives as a [GameEvent], possibly after other players' moves. A client
     * that treated the acknowledgement as the result would be rendering its own guess.
     *
     * [action] is the game's verb and [args] its operands, both game-defined. Pass [actionId] only to
     * retry a submission whose answer was lost; leaving it null mints a fresh one.
     */
    suspend fun submit(
        gameId: Id,
        roomId: Id,
        action: String,
        args: List<String>? = null,
        actionId: Long? = null,
    ): Acknowledged {
        val id = actionId ?: allocate(gameId)
        val request = GameAction(gameId, roomId, id, action, args)
        return rpc.call(
            Op.GAME_ACTION,
            { w -> request.encode(w) },
            { r -> Acknowledged.decode(r) },
        )
    }

    /**
     * The action id this client last used for a game, or null if it has submitted none.
     *
     * For a client that wants to persist its counter across a restart, and for a test that wants to
     * assert the sequence. Reading it does not advance it.
     */
    fun lastActionId(gameId: Id): Long? = nextActionId[gameId.value]?.minus(1L)

    /**
     * Forgets a game's action-id counter.
     *
     * Called when a game ends or a room is left. Keeping the counter would leak a slot per game played
     * for the lifetime of the process, and the server scopes ids to a game session anyway, so there is
     * nothing to preserve.
     */
    fun forget(gameId: Id) {
        nextActionId.remove(gameId.value)
    }

    /** Hands out the next action id for a game, atomically with respect to that game's key. */
    private fun allocate(gameId: Id): Long {
        var assigned = 1L
        nextActionId.compute(gameId.value) { _, current ->
            assigned = current ?: 1L
            assigned + 1L
        }
        return assigned
    }
}
